<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>BiocPy Packages – BiocPy: Facilitate Bioconductor Workflows in Python</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../index.html" rel="prev">
<link href=".././assets/short.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-b4278de3433e8877fdb1ba16cb94d3ca.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../custom.css">
<meta name="twitter:title" content="BiocPy Packages – BiocPy: Facilitate Bioconductor Workflows in Python">
<meta name="twitter:description" content="">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../sections/packages.html"><span class="chapter-title">BiocPy Packages</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">BiocPy: Facilitate Bioconductor Workflows in Python</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/BiocPy/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <a href="https://twitter.com/intent/tweet?url=|url|" title="Twitter" class="quarto-navigation-tool px-1" aria-label="Twitter"><i class="bi bi-twitter"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link"><span class="chapter-title">Welcome to BiocPy</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../sections/packages.html" class="sidebar-item-text sidebar-link active"><span class="chapter-title">BiocPy Packages</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#getting-started" id="toc-getting-started" class="nav-link active" data-scroll-target="#getting-started">Getting Started</a>
  <ul class="collapse">
  <li><a href="#installation" id="toc-installation" class="nav-link" data-scroll-target="#installation">Installation</a></li>
  </ul></li>
  <li><a href="#core-representations" id="toc-core-representations" class="nav-link" data-scroll-target="#core-representations">Core Representations</a></li>
  <li><a href="#analysis-packages" id="toc-analysis-packages" class="nav-link" data-scroll-target="#analysis-packages">Analysis Packages</a></li>
  <li><a href="#r-interoperability" id="toc-r-interoperability" class="nav-link" data-scroll-target="#r-interoperability">R Interoperability</a></li>
  <li><a href="#developer-guide" id="toc-developer-guide" class="nav-link" data-scroll-target="#developer-guide">Developer Guide</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-title">BiocPy Packages</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="getting-started" class="level2">
<h2 class="anchored" data-anchor-id="getting-started">Getting Started</h2>
<p>The BiocPy ecosystem is modular. You can install the collection of core packages via PyPI.</p>
<section id="installation" class="level3">
<h3 class="anchored" data-anchor-id="installation">Installation</h3>
<p>To get the core data structures and utilities:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install biocutils genomicranges summarizedexperiment singlecellexperiment multiassayexperiment</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>To install interoperability tools:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install rds2py txdb orgdb experimenthub</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
</section>
<section id="core-representations" class="level2">
<h2 class="anchored" data-anchor-id="core-representations">Core Representations</h2>
<p>These structures serve as essential and foundational data structures, acting as the building blocks for extensive and complex representations.</p>
<ul>
<li><strong><a href="https://github.com/biocpy/biocframe">BiocFrame</a></strong>: A Bioconductor-like data frame class.</li>
<li><strong><a href="https://github.com/biocpy/genomicranges">GenomicRanges</a></strong>: Aids in representing genomic regions and facilitating analysis.</li>
</ul>
<p>Container classes represent single or multi-omic experimental data and metadata:</p>
<ul>
<li><strong><a href="https://github.com/biocpy/summarizedexperiment">SummarizedExperiment</a></strong>: Container class to represent genomic experiments.</li>
<li><strong><a href="https://github.com/biocpy/singlecellexperiment">SingleCellExperiment</a></strong>: Container class to represent single-cell experiments.</li>
<li><strong><a href="https://github.com/biocpy/multiassayexperiment">MultiAssayExperiment</a></strong>: Container class to represent multiple experiments and assays performed over a set of samples.</li>
</ul>
<p>Moreover, BiocPy introduces a diverse range of data type classes designed to support the representation of atomic entities, including float, string, int lists, and named lists. These generics and utilities are provided through <a href="https://github.com/biocpy/biocutils">BiocUtils</a> package, and the delayed and file-backed array operations in the <a href="https://github.com/biocpy/delayedarray">DelayedArray</a> and their derivatives (<a href="https://github.com/biocpy/hdf5array">HDF5Array</a>, <a href="https://github.com/biocpy/tiledbarray">TileDbArray</a>).</p>
</section>
<section id="analysis-packages" class="level2">
<h2 class="anchored" data-anchor-id="analysis-packages">Analysis Packages</h2>
<p>BiocPy provides bindings to libscran and various other single-cell analysis methods incorporated into the <strong><a href="sections/workflows/scranpy.qmd">scranpy</a></strong> package to support analysis of multi-modal single-cell datasets. It also features integration with the <strong><a href="sections/workflows/singler.qmd">singler</a></strong> algorithm to annotate cell types by matching cells to known references based on their expression profiles.</p>
</section>
<section id="r-interoperability" class="level2">
<h2 class="anchored" data-anchor-id="r-interoperability">R Interoperability</h2>
<p>The <strong><a href="sections/interoperability/rds2py.qmd">rds2py</a></strong> package provides bindings to the rds2cpp library. This enables direct reading of RDS files in Python, eliminating the requirement for additional data conversion tools or intermediate formats. The package’s functionality streamlines the transition between Python and R, facilitating seamless analysis.</p>
<p>The following table serves as a directory of the core packages in the <strong>BiocPy</strong> ecosystem. All packages within the BiocPy ecosystem are published to <a href="https://pypi.org/org/BiocPy/">Python’s Package Index (PyPI)</a>.</p>
<div id="generate-table" class="cell" data-execution_count="2">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Package</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">Latest Version</th>
<th style="text-align: left;">Links</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>BiocFrame</strong></td>
<td style="text-align: left;">Flexible dataframe representation to support nested structures.</td>
<td style="text-align: left;"><code>0.7.2</code></td>
<td style="text-align: left;"><a href="https://github.com/BiocPy/biocframe">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>DelayedArray</strong></td>
<td style="text-align: left;">Delayed array operations from Bioconductor</td>
<td style="text-align: left;"><code>0.6.2</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/DelayedArray">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>GenomicRanges</strong></td>
<td style="text-align: left;">Container class to represent and operate over genomic regions and annotations.</td>
<td style="text-align: left;"><code>0.8.3</code></td>
<td style="text-align: left;"><a href="https://github.com/BiocPy/GenomicRanges">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>IRanges</strong></td>
<td style="text-align: left;">Python implementation of the <a href="https://bioconductor.org/packages/IRanges"><strong>IRanges</strong></a> Bioconductor package.</td>
<td style="text-align: left;"><code>0.7.1</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/iranges">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>MultiAssayExperiment</strong></td>
<td style="text-align: left;">Container class for representing and managing multi-omics genomic experiments</td>
<td style="text-align: left;"><code>0.6.0</code></td>
<td style="text-align: left;"><a href="https://github.com/BiocPy/MultiAssayExperiment">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>SingleCellExperiment</strong></td>
<td style="text-align: left;">Container class for single-cell experiments</td>
<td style="text-align: left;"><code>0.6.2</code></td>
<td style="text-align: left;"><a href="https://github.com/BiocPy/SingleCellExperiment">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>SpatialExperiment</strong></td>
<td style="text-align: left;">Container class for storing data from spatial-omics experiments</td>
<td style="text-align: left;"><code>0.1.0</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/spatialexperiment">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>SpatialFeatureExperiment</strong></td>
<td style="text-align: left;">Container class for storing data from spatial feature experiments</td>
<td style="text-align: left;"><code>0.0.5</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/spatialfeatureexperiment">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>SummarizedExperiment</strong></td>
<td style="text-align: left;">Container to represent data from genomic experiments</td>
<td style="text-align: left;"><code>0.6.5</code></td>
<td style="text-align: left;"><a href="https://github.com/BiocPy/summarizedexperiment">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>biocsetup</strong></td>
<td style="text-align: left;">“A CLI interface to quickly scaffold new BiocPy Python packages”</td>
<td style="text-align: left;"><code>0.3.3</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/biocsetup">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>biocutils</strong></td>
<td style="text-align: left;">Utilities to use across the biocpy packages.</td>
<td style="text-align: left;"><code>0.3.3</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/biocutils">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>biostrings</strong></td>
<td style="text-align: left;">Efficient manipulation of genomic sequences</td>
<td style="text-align: left;"><code>0.0.2</code></td>
<td style="text-align: left;"><a href="https://github.com/BiocPy/biostrings">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>celldex</strong></td>
<td style="text-align: left;">Index of Reference Cell Type Datasets</td>
<td style="text-align: left;"><code>0.3.0</code></td>
<td style="text-align: left;"><a href="https://github.com/SingleR-inc/celldex">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>compressed-lists</strong></td>
<td style="text-align: left;">Memory-efficient container for list-like objects</td>
<td style="text-align: left;"><code>0.4.4</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/compressed-lists">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>experimenthub</strong></td>
<td style="text-align: left;">Access Bioconductors experimenthub resources</td>
<td style="text-align: left;"><code>0.0.1</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/experimenthub">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>hdf5array</strong></td>
<td style="text-align: left;">HDF5-backed objects for array and matrix like data</td>
<td style="text-align: left;"><code>0.5.0</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/hdf5array">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>mopsy</strong></td>
<td style="text-align: left;">Matrix operations for numpy and scipy</td>
<td style="text-align: left;"><code>0.3.0</code></td>
<td style="text-align: left;"><a href="https://github.com/jkanche/mopsy">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>orgdb</strong></td>
<td style="text-align: left;">Access OrgDB annotations</td>
<td style="text-align: left;"><code>0.0.1</code></td>
<td style="text-align: left;"><a href="https://github.com/BiocPy/orgdb">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>pyBiocFileCache</strong></td>
<td style="text-align: left;">File based cache for resources and metadata</td>
<td style="text-align: left;"><code>0.7.0</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/pyBiocFileCache">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>rds2py</strong></td>
<td style="text-align: left;">Parse and construct Python representations for datasets stored in RDS files</td>
<td style="text-align: left;"><code>0.8.0</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/rds2py">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>scranpy</strong></td>
<td style="text-align: left;">Analyze multi-modal single-cell data!</td>
<td style="text-align: left;"><code>0.3.0</code></td>
<td style="text-align: left;"><a href="https://github.com/libscran/scranpy">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>scrnaseq</strong></td>
<td style="text-align: left;">Collection of Public Single-Cell RNA-Seq Datasets</td>
<td style="text-align: left;"><code>0.3.1</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/scrnaseq">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>singler</strong></td>
<td style="text-align: left;">Python bindings to the singleR algorithm to annotate cell types from known references.</td>
<td style="text-align: left;"><code>0.5.0</code></td>
<td style="text-align: left;"><a href="https://github.com/BiocPy/singler">GitHub</a></td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>tiledbarray</strong></td>
<td style="text-align: left;">TileDb backed objects for array and matrix like data</td>
<td style="text-align: left;"><code>0.2.0</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/tiledbarray/">GitHub</a></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>txdb</strong></td>
<td style="text-align: left;">Python interface to access and manipulate genome annotations in TxDB SQLite format.</td>
<td style="text-align: left;"><code>0.0.4</code></td>
<td style="text-align: left;"><a href="https://github.com/biocpy/txdb">GitHub</a></td>
</tr>
</tbody>
</table>
</div>
<hr>
</section>
<section id="developer-guide" class="level2">
<h2 class="anchored" data-anchor-id="developer-guide">Developer Guide</h2>
<p>If you are interested in developing new Python packages, check out the <a href="https://github.com/BiocPy/developer_guide">developer guide</a> on the phisolophy and tools we employ to ensure code quality and consistency within and across all the packages. A more detailed Python package management process is documented in the <a href="https://biocpy.github.io/BiocSetup/tutorial.html">biocsetup package</a>.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/biocpy\.github\.io\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../index.html" class="pagination-link" aria-label="Welcome to BiocPy">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-title">Welcome to BiocPy</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../chapters/summary.qmd">
<ol start="3" type="a">
<li>BiocPy core contributors</li>
</ol>
</a>
  </li>  
</ul>
    </div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>